### PR TITLE
Remove lambda from potential signatures

### PR DESCRIPTION
**Pull request by mcwitt**

### CODE REVIEW

File: tests/test_md.py [hunk 1]
```diff
@@ -39,35 +39,29 @@ def test_multiple_steps_store_interval(self):
...
         assert len(test_xs) == 1
-        assert len(test_du_dls) == 1
         assert len(test_boxes) == 1
 
         np.testing.assert_array_equal(test_xs[0], ctxt.get_x_t())
-        _, test_frame_du_dl, _ = bps[0].execute(test_xs[0], test_boxes[0], 0.0)
-        np.testing.assert_array_equal(test_du_dls[0], test_frame_du_dl)
+        _, _ = bps[0].execute(test_xs[0], test_boxes[0])
 
-    @pytest.mark.skip("multiple_steps_U is deprecated")
```
Comment (mcwitt) [inline tests/test_md.py:70]: This was mistakenly skipped in #893. Re-enable and update for the new `multiple_steps_U` interface.

File: tests/test_md.py [hunk 2]
```diff
@@ -314,17 +272,11 @@ def compute_reference_values():
...
-        test_us, test_xs, test_boxes = ctxt_3.multiple_steps_U(lamb, num_steps, lambda_windows, u_interval, x_interval)
-        np.testing.assert_array_almost_equal(ref_all_lambda_us[u_interval::u_interval], test_us)
+        test_us, test_xs, test_boxes = ctxt_3.multiple_steps_U(num_steps, u_interval, x_interval)
+        np.testing.assert_array_almost_equal(ref_all_us[u_interval::u_interval], test_us)
 
         np.testing.assert_array_almost_equal(ref_all_xs[x_interval::x_interval], test_xs)
 
-        test_us, test_xs, test_boxes = ctxt_3.multiple_steps_U(
-            lamb, num_steps, np.array([], dtype=np.float64), u_interval, x_interval
-        )
```
Comment (mcwitt) [inline tests/test_md.py:324]: This is no longer testing a corner case with the removal of the `lambda_schedule` argument.

File: timemachine/cpp/src/barostat.hpp [hunk 3]
```diff
@@ -8,18 +8,6 @@
...
-//public:
-//
-//    virtual void inplace_move(
-//        double *d_x,
-//        double *d_box,
-//        const double lambda
-//    );
-//
-//};
-
```
Comment (mcwitt) [inline timemachine/cpp/src/barostat.hpp:22]: Didn't seem worth keeping this up to date now that we should be able to get signatures in the IDE via `custom_ops.pyi`

File: tests/test_free_energy.py [hunk 4]
```diff
@@ -156,20 +154,18 @@ def test_construct_differentiable_interface_fast():
...
     sys_params = [np.array(bp.params) for bp in potentials]
 
     for precision in [np.float32, np.float64]:
         U_ref = construct_differentiable_interface(potentials, precision)
         U = construct_differentiable_interface_fast(potentials, sys_params, precision)
-        args = (coords, sys_params, box, lam)
+        args = (coords, sys_params, box)
         np.testing.assert_array_equal(U(*args), U_ref(*args))
 
-        argnums = (0, 1, 3)
```
Comment (badisa) [inline tests/test_free_energy.py:170]: Looks like this went from testing `[du_dx, du_dp, du_dl]` to `[du_dx]` since it no longer has `argnums = (0, 1)`. Is it intended to remove the `du_dp`?

Comment (mcwitt) [reply tests/test_free_energy.py:170]: Good catch, this was accidental. Will fix.

File: tests/test_jax_nonbonded.py [hunk 5]
```diff
@@ -191,12 +184,12 @@ def generate_random_inputs(n_atoms, dim, instance_flags=difficult_instance_flags
...
         beta += rand()
 
-    args = (conf, params, box, lamb, charge_rescale_mask, lj_rescale_mask, beta, cutoff)
+    args = (conf, params, box, charge_rescale_mask, lj_rescale_mask, beta, cutoff)
 
     return args
 
 
-def compare_two_potentials(u_a: NonbondedFxn, u_b: NonbondedFxn, args: NonbondedArgs, differentiate_wrt=(0, 1, 3)):
+def compare_two_potentials(u_a: NonbondedFxn, u_b: NonbondedFxn, args: NonbondedArgs, differentiate_wrt=(0, 1, 2)):
```
[the remark this answers is not in the document]
Comment (mcwitt) [reply tests/test_jax_nonbonded.py:192]: Good catch, we were not differentiating wrt box before. I'm not sure there's harm in doing this, but since we don't support dU/dbox in the custom ops potentials it probably doesn't make sense to test this here.

File: tests/test_free_energy.py (hunk 4 above)
Comment (mcwitt) [reply tests/test_free_energy.py:170]: Fixed in f490d3c

File: tests/test_jax_nonbonded.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcwitt) [reply tests/test_jax_nonbonded.py]: Fixed in d67f3de

**Review by badisa — APPROVED** [review]
🚀

File: timemachine/potentials/chiral_restraints.py [hunk 6]
```diff
@@ -110,7 +110,7 @@ def chiral_atom_restraint(conf, params, box, lamb, idxs):
     return jnp.sum(U_chiral_atom_batch_all(conf, idxs, params)) if len(idxs) else 0.0
 
 
-def chiral_bond_restraint(conf, params, box, lamb, idxs, signs):
+def chiral_bond_restraint(conf, params, box, idxs, signs):
```
Comment (maxentile) [inline timemachine/potentials/chiral_restraints.py:113]: nit: delete lamb reference in docstrings, L107, 119

File: timemachine/potentials/generic.py [hunk 7]
```diff
@@ -15,8 +15,7 @@
 Conf = Array
 Params = Array
 Box = Array
-Lambda = float
-PotentialFxn = Callable[[Conf, Params, Box, Lambda], float]
+PotentialFxn = Callable[[Conf, Params, Box], float]
```
Comment (maxentile) [inline timemachine/potentials/generic.py:18]: nit completeness --- Also remove unused `lamb` references in `rmsd.py`?

File: timemachine/md/minimizer.py [hunk 8]
```diff
@@ -41,7 +41,7 @@ def bind_potentials(params_potential_pairs):
     return u_impls
```
Comment (maxentile) [inline timemachine/md/minimizer.py:43]: nit completeness -- also update `get_val_and_grad_fn` in this file?

File: timemachine/fe/rbfe.py [hunk 9]
```diff
@@ -24,7 +24,7 @@
 from timemachine.potentials import jax_utils
 
 
-def get_batch_U_fns(bps, lamb):
+def get_batch_U_fns(bps):
     # return a function that takes in coords, boxes, lambda
```
Comment (maxentile) [inline timemachine/fe/rbfe.py:28]: nit: delete ref from comment

File: timemachine/cpp/src/context.cu [hunk 10]
```diff
@@ -19,7 +19,6 @@ Context::Context(
     d_x_t_ = gpuErrchkCudaMallocAndCopy(x_0, N * 3);
     d_v_t_ = gpuErrchkCudaMallocAndCopy(v_0, N * 3);
     d_box_t_ = gpuErrchkCudaMallocAndCopy(box_0, 3 * 3);
-    gpuErrchk(cudaMalloc(&d_du_dl_buffer_, N * sizeof(*d_du_dl_buffer_)));
     gpuErrchk(cudaMalloc(&d_u_buffer_, N * sizeof(*d_u_buffer_)));
 
     unsigned long long *d_in_tmp = nullptr;  // dummy
```
Comment (maxentile) [inline timemachine/cpp/src/context.cu:24]: nit : unsure -- comment on L27 references du_dl but I think describes code that's still needed regardless of du_dl?

File: tests/test_velocity_verlet_integrator.py [hunk 11]
```diff
@@ -107,13 +107,12 @@ def test_reversibility():
         # range of values in the velocities. Setting the seed
         # here keeps the range the same for all n_step values.
         np.random.seed(seed)
-        lamb_sched = np.linspace(0, 1, n_steps)
```
Comment (maxentile) [inline tests/test_velocity_verlet_integrator.py:110]: Observation (no action needed) -- This does weaken the property being tested, but that's fine -- I don't think we were relying on this property.

**Review by maxentile — APPROVED** [review]
🚀

I haven't read over _every_ line here, but the parts I've spot-checked look thorough and correct. Commented on a few non-blocking stragglers.

Remaining references I see are from files that are not being used / maintained:
* Can `md/ensembles.py` + its tests, `md/thermostat/*`, etc. be attic'd?
* I'll update references from SMC and reweighting in a later pass.

File: timemachine/cpp/src/context.cu (hunk 10 above)
Comment (mcwitt) [reply timemachine/cpp/src/context.cu:24]: Correct, we use `d_sum_storage_` and `d_sum_storage_bytes_` on L156 (in the reduction of per-atom energies). Will update the comment.

File: timemachine/potentials/chiral_restraints.py (hunk 6 above)
Comment (mcwitt) [reply timemachine/potentials/chiral_restraints.py:113]: Fixed in 370861d

File: timemachine/potentials/generic.py (hunk 7 above)
Comment (mcwitt) [reply timemachine/potentials/generic.py:18]: Removed in 004956e3b82be195357150b60dd8697ff9a2d91c

File: timemachine/md/minimizer.py (hunk 8 above)
Comment (mcwitt) [reply timemachine/md/minimizer.py:43]: Updated in e3aa8128dc5d17fba300f8c721ef5083f09852b1

File: timemachine/cpp/src/context.cu (hunk 10 above)
Comment (mcwitt) [reply timemachine/cpp/src/context.cu:24]: Updated comment in 440ba99

File: timemachine/fe/rbfe.py (hunk 9 above)
Comment (mcwitt) [reply timemachine/fe/rbfe.py]: Fixed in bfb226ccf18f4dbbd03de0e287df7bcac708b889

**Comment by mcwitt** [comment]
@maxentile 

> Commented on a few non-blocking stragglers.

Thanks, good catches.

> Can md/ensembles.py + its tests, md/thermostat/*, etc. be attic'd?

I don't see a reason not to; will go ahead with this if there are no objections.

> I'll update references from SMC and reweighting in a later pass.

Sounds good, thanks! :+1:

**Comment by mcwitt** [comment]
> Can md/ensembles.py + its tests, md/thermostat/*, etc. be attic'd?

Actually, it looks like `md/thermostat/utils.py` is still used in `test_barostat.py`. Maybe we should hold off on attic-ing for a later cleanup?

**Comment by maxentile** [comment]
>I don't see a reason not to; will go ahead with this if there are no objections.
>Actually, it looks like md/thermostat/utils.py is still used in test_barostat.py. Maybe we should hold off on attic-ing for a later cleanup?

Ahh, good call -- I think the one usage can be inlined, but I can do this in a later cleanup. (Only mentioned in the context of this PR since these are the only non-updated active references to `impl.execute(x, box, lam)`.)

Looks good to me regardless -- great simplification!

**Review by proteneer — APPROVED** [review]
this is pretty awesome
